src/ethereum_subscriber.rs

```diff
@@ -0,0 +1,388 @@
+        contracts: Arc<Contracts>,
+        tree_state: SharedTreeState,
+        identity_committer: Arc<IdentityCommitter>,
+    ) -> Self {
+        Self {
+            instance: RwLock::new(None),
+            starting_block,
+            database,
+            contracts,
+            tree_state,
+            identity_committer,
+        }
+    }
+
+    #[instrument(level = "debug", skip_all)]
+    pub async fn start(&self, refresh_rate: Duration) {
+        let mut instance = self.instance.write().await;
+        if instance.is_some() {
+            info!("Chain Subscriber already running");
+            return;
+        }
+
+        let mut starting_block = self.starting_block;
+        let database = self.database.clone();
+        let tree_state = self.tree_state.clone();
+        let contracts = self.contracts.clone();
+        let identity_committer = self.identity_committer.clone();
+
+        let handle = tokio::spawn(async move {
+            loop {
+                sleep(refresh_rate).await;
+
+                let processed_block = Self::process_events_internal(
+                    starting_block,
+                    tree_state.clone(),
+                    contracts.clone(),
+                    database.clone(),
+                    identity_committer.clone(),
+                )
+                .await;
+                match processed_block {
+                    Ok(block_number) => starting_block = block_number + 1,
+                    Err(error) => {
+                        error!(?error, "Couldn't process events update");
+                    }
+                }
+            }
+        });
+        *instance = Some(RunningInstance { handle });
+    }
+
+    #[instrument(level = "info", skip_all)]
+    pub async fn process_events(&mut self) -> Result<(), Error> {
+        let processed_block = Self::process_events_internal(
+            self.starting_block,
+            self.tree_state.clone(),
+            self.contracts.clone(),
+            self.database.clone(),
+            self.identity_committer.clone(),
+        )
+        .await?;
+        self.starting_block = processed_block + 1;
+        Ok(())
+    }
+
+    async fn process_events_internal(
+        starting_block: u64,
+        tree_state: SharedTreeState,
+        contracts: Arc<Contracts>,
+        database: Arc<Database>,
+        identity_committer: Arc<IdentityCommitter>,
+    ) -> Result<u64, Error> {
+        let end_block = contracts
+            .confirmed_block_number()
+            .await
+            .map_err(Error::EventError)?;
+
+        if starting_block > end_block {
+            return Ok(end_block);
+        }
+        info!(
+            starting_block,
+            end_block, "processing events in ethereum subscriber"
+        );
+
+        let mut events = contracts
+            .fetch_events(starting_block, Some(end_block), database.clone())
+            .boxed();
+
+        let shutdown = await_shutdown();
```

Comment:
I prefer to use a chan for this and pipe from `await_shutdown` explicitly – gives you more control in weird setups, like testing.

Reply:
Removed. Ethereum subscriber shouldn't care about system interrupts.